### DISCUSSION TITLE
Why can't/shouldn't you override a generated Notifier class with a mock of the interface during testing?

### DISCUSSION THREAD

**Discussion by danielmahon** [rrousselGit/riverpod#2095] [Q&A] [answered]
Here I expect to be able to override the repo inside the test with a mock of the "interface", not the implementation. This is possible through non-generated Providers. Has just been an inconvenience so far, but is there a way to tell the generator what the exposed class type should be?

```dart
// repository
@Riverpod(keepAlive: true)
class DeviceRepository extends _$DeviceRepository with Logger implements IDeviceRepository {
// ...
}

// tests
class MockDeviceRepository extends Mock implements IDeviceRepository {}

final container = ProviderContainer(overrides: [
  // Error: The return type 'MockDeviceRepository' isn't a 'DeviceRepository', as required by the closure's context.
  // Override overrideWith(DeviceRepository Function() create)
  deviceRepositoryProvider.overrideWith(() => mockDeviceRepository),
]);
```

**Comment by rrousselGit** [comment]
You can. You simply have a type mismatch in your scenario, which would happen with non-generated providers too

Your provider expects to return a DeviceRepository, yet your mock only implements IDeviceRepository not DeviceRepository

**Reply by danielmahon** [reply]
Thanks, yea I get that, but I don't see a way to get `deviceRepositoryProvider` to return (or expect) the `IDeviceRepository` interface instead. With non-gen providers I can give it the interface to return `final deviceRepositoryProvider = Provider< IDeviceRepository>(ref) => DeviceRepository();`

**Reply by rrousselGit** [reply]
Fair point. I'd suggest have your mock implement DeviceRepository instead of the interface

**Reply by danielmahon** [reply]
Cool. It will be fine for most use cases, though I could see it becoming more of a problem if you implement multiple interfaces on a class and only want/need to mock a single one.

**Reply by danielmahon** [reply]
It also looks like the function generator will let you return the interface and mock it. Just not the class annotation.
@rrousselGit, maybe there can be a way to add the return type to the annotation parameters? Or a generic on the base class? Although if you have multiple implementations of that interface, you would need to wrap it in a function or explicitly defined provider anyway to switch between them (edit: however, overriding them still wouldn't work because the interface wouldn't be exposed).

**Reply by rrousselGit** [reply]
That would be quite complex for very little.

**Reply by danielmahon** [reply]
Just to clarify, do you mean adding an annotation parameter or generic would be complex for little gain? Or the example use case? I can see lots of valid cases (I think) for needing to implement/override an interface multiple times. Either way not too big a deal, because you can just use the function generator:
```dart
@riverpod
IStorageRepository storageRepository(StorageRepositoryRef ref) {
  final isOffline = ref.watch(isOfflineProvider);
  return isOffline ? LocalStorageRepository() : RemoteStorageRepository();
}

//or override in widget tree (would not work with class annotation)
ProviderScope(overrides: [storageRepositoryProvider.overrideWithValue(LocalStorageRepository())],child:child);

class LocalStorageRepository implements IStorageRepository {
  //...
}

class RemoteStorageRepository implements IStorageRepository {
  //...
}
```

Thanks again for all the responses, I know you're still working on the docs, just wanted to ask all the questions so I don't miss anything important.

**Reply by AhmedLSayed9** [reply]
I'm using the function generator to be able to mock & override the interface with no problem.

The ability to define provider's type using the annotation `@riverpod(type: IRepo)` would be great too, unless It needs good effort to be implemented considering there's already an alternative?

Maybe it's better to open an enhancement issue or something.

**Comment by danielmahon** [comment, marked as answer]
Answering this as ~~you currently cannot return an interface from the `Notifier` generated using the class annotation~~ you can using a factory constructor and you can from the function generator. See https://github.com/rrousselGit/riverpod/discussions/2095#discussioncomment-4738580

**Comment by rrousselGit** [comment]
You could use a factory constructor on your notifier class:

```dart
@riverpod
class MyNotifier extends _$MyNotifier {
  factory MyNotifier() = _MyNotifierImpl;
  
  @override
  int build();
}

class _MyNotifierImpl extends MyNotifier {
  @override
  int build() {}
}
```


In this scenario, `MyNotifier` is the interface and `MyNotifierImpl` is the concrete implementation
